synapse/handlers/ui_auth/checkers.py

```diff
@@ -33,25 +33,25 @@ class UserInteractiveAuthChecker:
     def __init__(self, hs):
         pass
 
-    def is_enabled(self):
+    def is_enabled(self) -> bool:
         """Check if the configuration of the homeserver allows this checker to work
 
         Returns:
-            bool: True if this login type is enabled.
+            True if this login type is enabled.
         """
 
-    def check_auth(self, authdict, clientip):
+    async def check_auth(self, authdict: dict, clientip: str) -> Any:
```

Comment:
Were you holding off specifying the possibilities here so that you wouldn't need to specify `Deferred`? If so, it's worth noting that `isinstance(a_deferred, Awaitable)` resolves to `True`, so I *think* something not too cumbersome like `Optional[Union[str, Tuple]]` should work here.

Reply:
I didn't specify the possibilities because I did the typing before reading the separate documentation and the docstring didn't specify what was returned. 😄 I can double check the return types to the documentation.

Reply:
Oh wait, I know why I did this. (I should have ☕ before replying to things...)

Although this method is also named `check_auth`, it is *NOT* the same as above. It gets called twice:

https://github.com/matrix-org/synapse/blob/4e5ec32e022ed79628f8c231a9d0c682d7d8ac0f/synapse/handlers/auth.py#L426-L430

https://github.com/matrix-org/synapse/blob/4e5ec32e022ed79628f8c231a9d0c682d7d8ac0f/synapse/handlers/auth.py#L510-L513

While the one from a password provider gets called:

https://github.com/matrix-org/synapse/blob/4e5ec32e022ed79628f8c231a9d0c682d7d8ac0f/synapse/handlers/auth.py#L754-L758

Anyway, the result of `UserInteractiveAuthChecker.check_auth` gets saved into the database and sometimes inspected in other places.

Reply:
Ah, fair enough then! Thanks for the clear explanation :)

Reply:
You're welcome! I'm now realizing that putting these changes in the same PR was confusing...since they're not really related. 😢 Sorry about that!